pkg/webhook/pod/tikv_deleter.go

```diff
@@ -0,0 +1,208 @@
+	if err != nil {
+		klog.Infof("tc[%s/%s]'s tikv pod[%s/%s] failed to delete,%v", namespace, tcName, namespace, name, err)
+		return util.ARFail(err)
+	}
+
+	isUpgrading := IsStatefulSetUpgrading(ownerStatefulSet)
+
+	if storeInfo == nil || !isStoreExist {
+		return pc.admitDeleteNonStoreTiKVPod(isInOrdinal, pod, ownerStatefulSet, tc, pdClient)
+	}
+
+	if storeInfo.Store == nil {
+		return util.ARFail(err)
+	}
+
+	switch storeInfo.Store.StateName {
+	case v1alpha1.TiKVStateTombstone:
+		return pc.admitDeleteTombStoneTiKVPod(isInOrdinal, pod, ownerStatefulSet, tc, pdClient)
+	case v1alpha1.TiKVStateOffline:
+		return pc.admitDeleteOfflineTiKVPod()
+	case v1alpha1.TiKVStateDown:
+		return pc.admitDeleteDownTiKVPod(isInOrdinal, pod, ownerStatefulSet, tc, pdClient)
+	case v1alpha1.TiKVStateUp:
+		return pc.admitDeleteUpTiKVPod(isInOrdinal, isUpgrading, pod, ownerStatefulSet, tc, pdClient, storeInfo, storesInfo)
+	}
+
+	return &admission.AdmissionResponse{
+		Allowed: false,
+	}
+}
+
+func (pc *PodAdmissionControl) admitDeleteNonStoreTiKVPod(isInOrdinal bool, pod *core.Pod, ownerStatefulSet *apps.StatefulSet, tc *v1alpha1.TidbCluster, pdClient pdapi.PDClient) *admission.AdmissionResponse {
+
+	name := pod.Name
+	namespace := pod.Namespace
+	tcName := tc.Name
+	ordinal, err := operatorUtils.GetOrdinalFromPodName(name)
+	if err != nil {
+		return util.ARFail(err)
+	}
+	setName := ownerStatefulSet.Name
+
+	if !isInOrdinal {
+		err = addDeferDeletingToPVC(v1alpha1.TiKVMemberType, pc, tc, setName, namespace, ordinal)
+		if err != nil {
+			klog.Infof("tc[%s/%s]'s tikv pod[%s/%s] failed to delete,%v", namespace, tcName, namespace, name, err)
+			return util.ARFail(err)
+		}
+	}
+
+	return util.ARSuccess()
+}
+
+func (pc *PodAdmissionControl) admitDeleteTombStoneTiKVPod(isInOrdinal bool, pod *core.Pod, ownerStatefulSet *apps.StatefulSet, tc *v1alpha1.TidbCluster, pdClient pdapi.PDClient) *admission.AdmissionResponse {
+
+	name := pod.Name
+	namespace := pod.Namespace
+	tcName := tc.Name
+	ordinal, err := operatorUtils.GetOrdinalFromPodName(name)
+	if err != nil {
+		klog.Infof("tc[%s/%s]'s tikv pod[%s/%s] failed to delete,%v", namespace, tcName, namespace, name, err)
+		return util.ARFail(err)
+	}
+	setName := ownerStatefulSet.Name
+	if !isInOrdinal {
+		err = addDeferDeletingToPVC(v1alpha1.TiKVMemberType, pc, tc, setName, namespace, ordinal)
+		if err != nil {
+			klog.Infof("tc[%s/%s]'s tikv pod[%s/%s] failed to delete,%v", namespace, tcName, namespace, name, err)
+			return util.ARFail(err)
+		}
+	}
+	return util.ARSuccess()
+}
+
+func (pc *PodAdmissionControl) admitDeleteOfflineTiKVPod() *admission.AdmissionResponse {
+	return &admission.AdmissionResponse{
+		Allowed: false,
+	}
+}
+
+func (pc *PodAdmissionControl) admitDeleteDownTiKVPod(isInOrdinal bool, pod *core.Pod, ownerStatefulSet *apps.StatefulSet, tc *v1alpha1.TidbCluster, pdClient pdapi.PDClient) *admission.AdmissionResponse {
```

Comment:
The functions are the same for Down, NonStore, and Tombstone.

Reply:
> The functions are the same for Down, NonStore, and Tombstone.

Yep, separating into different functions is intended if they become different in future.